### PR TITLE
Include FieldName in Hash to avoid collision

### LINKED ISSUE #11: Hasher ignores field names which can cause collisions
By only including only the field value (and ignoring the field name) in the hasher, the following objects can collide:

```
collidableType {
  a string
  b string
}

first = collidableType {
  a: "value"
}

second = collidableType {
  b: "value"
}
```

Since properties `a` and `b` have the same type and value, when the hasher evaluates the `first` and `second` it will produce identical hashes, even though they are distinct.

### PR DESCRIPTION
**Pull request by sam-heilbron**

### CODE REVIEW

**Comment by sam-heilbron** [comment]
Anything stopping me from ripping out the gogo references in this repo as well?

File: tests/hash_test.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 
 var _ = Describe("hash", func() {
 	testRuns := 100
+
```
Comment (mlholland) [inline tests/hash_test.go]: Extra whitespace added here.

**Comment by mlholland** [comment]
> Anything stopping me from ripping out the gogo references in this repo as well?

Sounds like a question for @EItanya

**Comment by EItanya** [comment]
It should be fine to drop it at this point

File: tests/hash_test.go [hunk 2]
```diff
@@ -173,7 +174,57 @@ var _ = Describe("hash", func() {
...
+			}
+			for idx := 0; idx < testRuns; idx++ {
+				newHashes := make([]uint64, len(testCases))
+				for i, v := range testCases {
+					hash, err := v.Hash(nil)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(hash).To(Equal(hashes[i]))
+					newHashes[i] = hash
+				}
+				Expect(allDifferent(newHashes)).To(BeTrue())
```
Comment (npolshakova) [inline tests/hash_test.go:223]: 🎉

File: templates/hash/functions.go [hunk 3]
```diff
@@ -48,6 +48,10 @@ func (fns goSharedFuncs) pointerAccessor(field pgs.Field) string {
 	return fmt.Sprintf("&m.%s", fns.Name(field))
 }
 
+func (fns goSharedFuncs) fieldName(field pgs.Field) string {
+	return fmt.Sprintf("%s", fns.Name(field))
```
Comment (saiskee) [inline templates/hash/functions.go:52]: just curious, what is the advantage of doing this over `string(fns.Name(field))`?

**Review by EItanya — APPROVED** [review]

File: templates/hash/functions.go (hunk 3 above)
Comment (sam-heilbron) [reply templates/hash/functions.go:52]: No reason in particular. I was just following the format of the other functions. I probably could have just used `fns.Name(field).ToString()`

**Pull request merged**